app/models/katello/concerns/redhat_extensions.rb

```diff
@@ -44,16 +44,31 @@ def construct_name(family)
       end
 
       def medium_uri_with_content_uri(host, url = nil)
-        if host.try(:content_source) && (repo_details = kickstart_repo(host))
-          URI.parse(repo_details[:path])
+        if host.try(:content_source) && (repos = distribution_repositories(host)).present?
+          if repos.size == 1
+            URI.parse(repos.first.full_path(host.content_source))
+          else
+            uri = medium_uri_without_content_uri(host, url)
+            uri_path = uri.path[-1] == '/' ? uri.path : uri.path + "/"
+            distro_repo = repos.find do |repo|
+              path = repo.full_path(host.content_source)
+              path += "/" unless path[-1] == '/'
+              path.include?(uri_path)
+            end
+            URI.parse(distro_repo.full_path(host.content_source))
+          end
         else
           medium_uri_without_content_uri(host, url)
         end
       end
 
-      def kickstart_repo(host)
-        distro = distribution_repositories(host).first
-        {:name => distro.name, :path => distro.full_path(host.content_source)} if distro && host.content_source
+      def kickstart_repos(host)
+        distros = distribution_repositories(host)
+        if distros && host.content_source
+          distros.map do |distro|
+            {:name => distro.name, :path => distro.full_path(host.content_source)}
+          end
+        end
```

Comment:
Right now this would return nil if no distros or the host does not have a content source, should it return an empty array for consistency?